File: backend/src/test/kotlin/app/meals/IngredientLineParserTest.kt
package app.meals

import app.meals.importer.IngredientLineParser
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class IngredientLineParserTest {

    @Test
    fun `norwegian decimal comma and dl`() {
        val ing = IngredientLineParser.parseLine("1,5 dl fin sammalt hvete")
        assertEquals("1,5", ing.quantity)
        assertEquals("dl", ing.unit)
        assertEquals("fin sammalt hvete", ing.name)
    }

    @Test
    fun `norwegian tablespoon abbreviation`() {
        val ing = IngredientLineParser.parseLine("1 ss margarin til steking")
        assertEquals("1", ing.quantity)
        assertEquals("ss", ing.unit)
        assertEquals("margarin til steking", ing.name)
    }

    @Test
    fun `approximate count stk with dotted abbreviation`() {
        val ing = IngredientLineParser.parseLine("ca. 20 stk. maistortilla")
        assertEquals("20", ing.quantity)
        assertEquals("stk", ing.unit)
        assertEquals("ca. maistortilla", ing.name)
    }

    @Test
    fun `english style from json-ld fixtures`() {
        val ing = IngredientLineParser.parseLine("200 g carrots")
        assertEquals("200", ing.quantity)
        assertEquals("g", ing.unit)
        assertEquals("carrots", ing.name)
    }

    @Test
    fun `unrecognized unit keeps full line in name`() {
        val line = "a pinch of salt"
        val ing = IngredientLineParser.parseLine(line)
        assertEquals("", ing.quantity)
        assertEquals("", ing.unit)
        assertEquals(line, ing.name)
    }

    @Test
    fun `no trailing name falls back to full line`() {
        val line = "200 g"
        val ing = IngredientLineParser.parseLine(line)
        assertEquals("", ing.quantity)
        assertEquals("", ing.unit)
        assertEquals(line, ing.name)
    }
}

File: backend/src/importer/ingredientLineParser.test.ts
import { describe, expect, it } from "vitest";
import { parseLine } from "./ingredientLineParser.js";

describe("parseLine", () => {
  it("handles a Norwegian decimal comma and dl", () => {
    expect(parseLine("1,5 dl fin sammalt hvete")).toEqual({
      name: "fin sammalt hvete",
      quantity: "1,5",
      unit: "dl",
    });
  });

  it("handles the Norwegian tablespoon abbreviation", () => {
    expect(parseLine("1 ss margarin til steking")).toEqual({
      name: "margarin til steking",
      quantity: "1",
      unit: "ss",
    });
  });

  it("keeps an approximate qualifier on the name and normalizes a dotted count unit", () => {
    expect(parseLine("ca. 20 stk. maistortilla")).toEqual({
      name: "ca. maistortilla",
      quantity: "20",
      unit: "stk",
    });
  });

  it("parses an English-style line", () => {
    expect(parseLine("200 g carrots")).toEqual({ name: "carrots", quantity: "200", unit: "g" });
  });

  it("keeps the full line as name for an unrecognized unit", () => {
    expect(parseLine("a pinch of salt")).toEqual({
      name: "a pinch of salt",
      quantity: "",
      unit: "",
    });
  });

  it("falls back to the full line when there is no trailing name", () => {
    expect(parseLine("200 g")).toEqual({ name: "200 g", quantity: "", unit: "" });
  });

  it("returns an empty ingredient for blank input", () => {
    expect(parseLine("   ")).toEqual({ name: "", quantity: "", unit: "" });
  });
});
